transhumandesign/kag-base
# [fixed] floating Arrows

## Status

- **READY**: this PR is (to the best of your knowledge) ready to be incorporated into the game.

## Description

[removed] TeamStructureNear.as from Arrow.as as it seems unused.
[removed] A check in Arrows.as (see lines 168-177) because it seems useless.
[changed] isOpen() is the same between SwingDoor, TrapBlock and Bridge, so moved it to DoorCommon.as.
[fixed] Arrows floating in the air.

This is a simple floating arrows fix.

When an arrow hits a structure blob, then that blob is passed to ArrowHitMap(), 
otherwise if the map is hit then null is passed.

In the former case, a reference to the hit blob is saved to the arrow,
in the latter case, the endpos of the map raycast check is saved to the arrow.

onTick() which is called every tick deals with these two scenarios and checks if the structure blob is still there,
or if the structure blob or map tile is still solid. Otherwise the arrow is set to die.

Please note, this was tested only locally. This has yet to be tested on a dedicated server.

## Steps to Test or Reproduce

1. Shoot an arrow at a door. Walk through it. Notice it dies.
2. Shoot an arrow at a door/platform/bridge/magazine/etc. that is built on a wall. Make the structure collapse. Notice the arrow dies.
3. Shoot an arrow at a bridge, switch team and fall through it. Notice the arrow dies.
4. /loadmap Princess06, shoot an arrow at the trap blocks, fall through. Notice the arrow dies.
5. Build obstructor blocks and activate them via switch. Shoot at them. Then disable the blocks via switch. Notice the arrow dies.

## How to make this Better?

I said this is a simple fix. I think these would be nice to have. Not sure if I will work on it:
- When the door with arrow on it is opened, make the arrow shake and fall only after 1-2 sec. So enemy has to leave himself vulnerable.
- Instead of simply dying, it would be nice if the arrows fell down (kind of like a second shot).
- I made a PR [here](https://github.com/transhumandesign/kag-base/pull/1249) which makes the Dummy strawman hittable by arrow. The arrow will be stuck on it. So just in case both PRs get accepted, the behavior of Arrow sticking to a Dummy should be looked at.
- Instead of dying when hitting an enemy crate or catapult, make it stick, move and rotate along with those blobs.
